bundles/org.openhab.binding.speedtest/src/main/java/org/openhab/binding/speedtest/internal/SpeedtestHandler.java

```diff
@@ -0,0 +1,525 @@
+
+        newState = new QuantityType<>(Double.parseDouble(ping_latency) / 1000.0, Units.SECOND);
+        logger.debug("ping_latency: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.PING_LATENCY), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(download_bandwidth) / 125000.0, Units.MEGABIT_PER_SECOND);
+        logger.debug("download_bandwidth: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.DOWNLOAD_BANDWIDTH), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(download_bytes), Units.BYTE);
+        logger.debug("download_bytes: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.DOWNLOAD_BYTES), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(download_elapsed) / 1000.0, Units.SECOND);
+        logger.debug("download_elapsed: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.DOWNLOAD_ELAPSED), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(upload_bandwidth) / 125000.0, Units.MEGABIT_PER_SECOND);
+        logger.debug("upload_bandwidth: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.UPLOAD_BANDWIDTH), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(upload_bytes), Units.BYTE);
+        logger.debug("upload_bytes: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.UPLOAD_BYTES), newState);
+
+        newState = new QuantityType<>(Double.parseDouble(upload_elapsed) / 1000.0, Units.SECOND);
+        logger.debug("upload_elapsed: {}", newState);
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.UPLOAD_ELAPSED), newState);
+
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.INTERFACE_EXTERNALIP),
+                new StringType(String.valueOf(interface_externalIp)));
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.INTERFACE_INTERNALIP),
+                new StringType(String.valueOf(interface_internalIp)));
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.ISP),
+                new StringType(String.valueOf(isp)));
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.RESULT_URL),
+                new StringType(String.valueOf(result_url)));
+        updateState(new ChannelUID(getThing().getUID(), SpeedtestBindingConstants.SERVER),
+                new StringType(String.valueOf(server_name)));
+    }
+
+    /*
+     * Checks to make sure the executable for speedtest is valid
+     */
+    public boolean checkConfig(String execPath) {
+        if (checkExecPath(execPath)) { // Check if entered path exists
```

Comment:
While the path might exists, the user that runs openhab might not have permissions to execute a command, or have permissions to execute this file. These permission issuesw might occur or not? If so, it would be good to update the bridge state with a helpfull text.


Maybe you can check by doing some speedtestcli.exe -v or similar version check. to test all permissions are setup nicely.

Reply:
Added `file.canExecute()` to the configuration check:
```
    private static boolean fileExistsAndIsExecutable(File file) {
        return file.exists() && !file.isDirectory() && file.canExecute();
    }
```

> Maybe you can check by doing some speedtestcli.exe -v

This is already implemented some lines below via `getSpeedTestVersion()`:
```
        if (!checkConfig(speedTestCommand)) { // check the config
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                    "Speedtest executable not found or not executable. Please check configuration.");
            return;
        }
        if (!getSpeedTestVersion()) {
            return;
        }
```

Reply:
@lsiepel, your findings seem to be addressed. Please have a look again.

Reply:
Great that this check is added. Minor thought i had was that it might be good to seperate the checks if the file is available or executable, as that might result in a more specific error to the user, but that is really a minor thought.

Reply:
Done. Split between `checkFileExists()` and `checkFileExecutable()`.